OllamaStreamingChatLanguageModel should support tools

Currently the `OllamaStreamingChatLanguageModel` doesn't support tool integration. Let's add it.

I'll take this and work on it.

That makes sense. I actually just tried it with OpenAI...

```
2024-07-30 14:58:14,641 INFO  [io.qua.lan.ope.OpenAiRestApi$OpenAiClientLogger] (vert.x-eventloop-thread-1) Request:
- method: POST
- url: https://parasol-chat-aiworkshop.apps.cluster-wk5tx.sandbox3051.opentlc.com/v1/chat/completions
- headers: [Accept: text/event-stream], [Authorization: Be...my], [Content-Type: application/json], [User-Agent: langchain4j-openai], [content-length: 2538]
- body: {
  "model" : "parasol-chat",
  "messages" : [ {
    "role" : "system",
    "content" : "You are a helpful, respectful and honest assistant named \"Parasol Assistant\".\nYou will be given a claim summary, references to provide you with information, and a question. You must answer the question based as much as possible on this claim with the help of the references.\nAlways answer as helpfully as possible, while being safe. Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\n\nIf a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\n"
  }, {
    "role" : "user",
    "content" : "    Claim ID: 1\n\n    Claim Summary:\n    \n    On January 2nd, 1955, at around 3:30 PM, a car accident occurred at the intersection of Colima Road and Azusa Avenue in Hill Vallet. The involved parties were Marty McFly, driving a silver Delorean DMC-12 (OUTA-TIME), and Biff Tanner in a blue Type 2 Volkswagen Bus (BIF-RULZ).\n\n    Marty was heading south on Colima Road when Biff failed to stop at the red traffic signal on Asuza Avenue, causing a collision with Marty's vehicle. Both drivers exchanged information and took photos of the accident scene, which included damages to the front driver and passenger side of Marty's Delorean DMC-12 and the front driver's side of Biff's Volkswagen Bus. No injuries were reported.\n\n    Marty has attached necessary documents, such as photos, a police report, and an estimate for repair costs, to his email. He requests prompt attention to the claim and is available at (916) 555-4385 or marty.mcfly@email.com for any additional information or documentation needed.\n    \n\n    Question: Should I approve this claim?\n"
  } ],
  "temperature" : 0.3,
  "top_p" : 1.0,
  "stream" : true,
  "stop" : [ "DONE", "done", "stop", "STOP" ],
  "presence_penalty" : 0.0,
  "frequency_penalty" : 0.0,
  "tools" : [ {
    "type" : "function",
    "function" : {
      "name" : "updateClaimStatus",
      "description" : "update claim status",
      "parameters" : {
        "type" : "object",
        "properties" : {
          "claimId" : {
            "type" : "integer"
          },
          "status" : {
            "type" : "string"
          }
        },
        "required" : [ "claimId", "status" ]
      }
    }
  } ]
}

2024-07-30 14:58:14,702 ERROR [io.qua.mut.run.MutinyInfrastructure] (vert.x-eventloop-thread-1) Mutiny had to drop the following exception: io.smallrye.mutiny.CompositeException: Multiple exceptions caught:
        [Exception 0] dev.ai4j.openai4j.OpenAiHttpException: {"object":"error","message":"[{'type': 'extra_forbidden', 'loc': ('body', 'tools'), 'msg': 'Extra inputs are not permitted', 'input': [{'type': 'function', 'function': {'name': 'updateClaimStatus', 'description': 'update claim status', 'parameters': {'type': 'object', 'properties': {'claimId': {'type': 'integer'}, 'status': {'type': 'string'}}, 'required': ['claimId', 'status']}}}]}]","type":"BadRequestError","param":null,"code":400}
        [Exception 1] java.lang.NullPointerException: Cannot invoke "dev.langchain4j.model.output.Response.content()" because "response" is null
        at io.smallrye.mutiny.subscription.Subscribers$CallbackBasedSubscriber.onFailure(Subscribers.java:95)
        at io.smallrye.mutiny.operators.multi.builders.BaseMultiEmitter.failed(BaseMultiEmitter.java:89)
        at io.smallrye.mutiny.operators.multi.builders.BufferItemMultiEmitter.drain(BufferItemMultiEmitter.java:108)
        at io.smallrye.mutiny.operators.multi.builders.BufferItemMultiEmitter.failed(BufferItemMultiEmitter.java:62)
        at io.smallrye.mutiny.operators.multi.builders.BaseMultiEmitter.fail(BaseMultiEmitter.java:78)
        at io.smallrye.mutiny.operators.multi.builders.SerializedMultiEmitter.drainLoop(SerializedMultiEmitter.java:110)
        at io.smallrye.mutiny.operators.multi.builders.SerializedMultiEmitter.drain(SerializedMultiEmitter.java:92)
        at io.smallrye.mutiny.operators.multi.builders.SerializedMultiEmitter.onFailure(SerializedMultiEmitter.java:77)
        at io.smallrye.mutiny.operators.multi.builders.SerializedMultiEmitter.fail(SerializedMultiEmitter.java:149)
        at org.jboss.resteasy.reactive.client.impl.MultiInvoker.lambda$method$0(MultiInvoker.java:127)
        at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
        at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.handleUnrecoverableError(RestClientRequestContext.java:368)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:329)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:175)
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:314)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:279)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:261)
        at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:59)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        Suppressed: java.lang.NullPointerException: Cannot invoke "dev.langchain4j.model.output.Response.content()" because "response" is null
                at dev.langchain4j.model.openai.InternalOpenAiHelper.removeTokenUsage(InternalOpenAiHelper.java:283)
                at dev.langchain4j.model.openai.OpenAiStreamingChatModel.createResponse(OpenAiStreamingChatModel.java:251)
                at dev.langchain4j.model.openai.OpenAiStreamingChatModel.lambda$generate$5(OpenAiStreamingChatModel.java:217)
                at io.smallrye.mutiny.subscription.Subscribers$CallbackBasedSubscriber.onFailure(Subscribers.java:93)
                ... 28 more
Caused by: dev.ai4j.openai4j.OpenAiHttpException: {"object":"error","message":"[{'type': 'extra_forbidden', 'loc': ('body', 'tools'), 'msg': 'Extra inputs are not permitted', 'input': [{'type': 'function', 'function': {'name': 'updateClaimStatus', 'description': 'update claim status', 'parameters': {'type': 'object', 'properties': {'claimId': {'type': 'integer'}, 'status': {'type': 'string'}}, 'required': ['claimId', 'status']}}}]}]","type":"BadRequestError","param":null,"code":400}
        at io.quarkiverse.langchain4j.openai.OpenAiRestApi.toException(OpenAiRestApi.java:170)
        at io.quarkiverse.langchain4j.openai.OpenAiRestApi_toException_ResponseExceptionMapper_f35c1c86580504f69920f9de921a22bd696c020f.toThrowable(Unknown Source)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientResponseFilter.filter(MicroProfileRestClientResponseFilter.java:49)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:21)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:10)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        ... 12 more
```

--- a/model-providers/ollama/deployment/src/test/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaStreamingChatLanguageModelSmokeTest.java
+++ b/model-providers/ollama/deployment/src/test/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaStreamingChatLanguageModelSmokeTest.java
@@ -1,0 +1,281 @@
+package io.quarkiverse.langchain4j.ollama.deployment;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.equalToJson;
+import static com.github.tomakehurst.wiremock.client.WireMock.post;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.util.List;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.github.tomakehurst.wiremock.stubbing.Scenario;
+
+import dev.langchain4j.agent.tool.Tool;
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.data.message.ToolExecutionResultMessage;
+import dev.langchain4j.service.MemoryId;
+import dev.langchain4j.store.memory.chat.ChatMemoryStore;
+import io.quarkiverse.langchain4j.RegisterAiService;
+import io.quarkiverse.langchain4j.testing.internal.WiremockAware;
+import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Multi;
+
+public class OllamaStreamingChatLanguageModelSmokeTest extends WiremockAware {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(Calculator.class))
+            .overrideConfigKey("quarkus.langchain4j.ollama.base-url", WiremockAware.wiremockUrlForConfig())
+            .overrideConfigKey("quarkus.langchain4j.devservices.enabled", "false");
+
+    @Singleton
+    @RegisterAiService(tools = Calculator.class)
+    interface AIServiceWithTool {
+        Multi<String> streaming(@MemoryId String memoryId, @dev.langchain4j.service.UserMessage String text);
+    }
+
+    @Singleton
+    @RegisterAiService
+    interface AIServiceWithoutTool {
+        Multi<String> streaming(@dev.langchain4j.service.UserMessage String text);
+    }
+
+    @Singleton
+    static class Calculator {
+        @Tool("Execute the sum of two numbers")
+        public int sum(int firstNumber, int secondNumber) {
+            return firstNumber + secondNumber;
+        }
+    }
+
+    @Inject
+    AIServiceWithTool aiServiceWithTool;
+
+    @Inject
+    AIServiceWithoutTool aiServiceWithoutTool;
+
+    @Inject
+    ChatMemoryStore memory;
+
+    @Test
+    void test_1() {
+        wiremock().register(
+                post(urlEqualTo("/api/chat"))
+                        .withRequestBody(equalToJson("""
+                                {
+                                  "model" : "llama3.2",
+                                  "messages" : [ {
+                                    "role" : "user",
+                                    "content" : "Hello"
+                                  }],
+                                  "options" : {
+                                    "temperature" : 0.8,
+                                    "top_k" : 40,
+                                    "top_p" : 0.9
+                                  },
+                                  "stream" : true
+                                }
+                                """))
+                        .willReturn(aResponse()
+                                .withHeader("Content-Type", "application/x-ndjson")
+                                .withBody(
+                                        """
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:42.312611426Z","message":{"role":"assistant","content":"Hello"},"done":false}
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:42.514215351Z","message":{"role":"assistant","content":"!"},"done":false}
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:44.109059873Z","message":{"role":"assistant","content":""},"done_reason":"stop","done":true,"total_duration":4821417857,"load_duration":2508844071,"prompt_eval_count":11,"prompt_eval_duration":514000000,"eval_count":10,"eval_duration":1797000000}""")));
+
+        var result = aiServiceWithoutTool.streaming("Hello").collect().asList().await().indefinitely();
+        assertEquals(List.of("Hello", "!"), result);
+    }
+
+    @Test
+    void test_2() {
+        wiremock().register(
+                post(urlEqualTo("/api/chat"))
+                        .withRequestBody(equalToJson("""
+                                {
+                                  "model" : "llama3.2",
+                                  "messages" : [ {
+                                    "role" : "user",
+                                    "content" : "Hello"
+                                  }],
+                                  "tools" : [ {
+                                    "type" : "function",
+                                    "function" : {
+                                      "name" : "sum",
+                                      "description" : "Execute the sum of two numbers",
+                                      "parameters" : {
+                                        "type" : "object",
+                                        "properties" : {
+                                          "firstNumber" : {
+                                            "type" : "integer"
+                                          },
+                                          "secondNumber" : {
+                                            "type" : "integer"
+                                          }
+                                        },
+                                        "required" : [ "firstNumber", "secondNumber" ]
+                                      }
+                                    }
+                                  } ],
+                                  "options" : {
+                                    "temperature" : 0.8,
+                                    "top_k" : 40,
+                                    "top_p" : 0.9
+                                  },
+                                  "stream" : true
+                                }
+                                """))
+                        .willReturn(aResponse()
+                                .withHeader("Content-Type", "application/x-ndjson")
+                                .withBody(
+                                        """
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:42.312611426Z","message":{"role":"assistant","content":"Hello"},"done":false}
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:42.514215351Z","message":{"role":"assistant","content":"!"},"done":false}
+                                                {"model":"llama3.2","created_at":"2024-11-30T09:03:44.109059873Z","message":{"role":"assistant","content":""},"done_reason":"stop","done":true,"total_duration":4821417857,"load_duration":2508844071,"prompt_eval_count":11,"prompt_eval_duration":514000000,"eval_count":10,"eval_duration":1797000000}""")));
+
+        var result = aiServiceWithTool.streaming("1", "Hello").collect().asList().await().indefinitely();
+        assertEquals(List.of("Hello", "!"), result);
+    }
+
+    @Test
+    void test_3() {
+        wiremock()
+                .register(
+                        post(urlEqualTo("/api/chat"))
+                                .inScenario("")
+                                .whenScenarioStateIs(Scenario.STARTED)
+                                .willSetStateTo("TOOL_CALL")
+                                .withRequestBody(equalToJson("""
+                                        {
+                                          "model" : "llama3.2",
+                                          "messages" : [ {
+                                            "role" : "user",
+                                            "content" : "1 + 1"
+                                          }],
+                                          "tools" : [ {
+                                            "type" : "function",
+                                            "function" : {
+                                              "name" : "sum",
+                                              "description" : "Execute the sum of two numbers",
+                                              "parameters" : {
+                                                "type" : "object",
+                                                "properties" : {
+                                                  "firstNumber" : {
+                                                    "type" : "integer"
+                                                  },
+                                                  "secondNumber" : {
+                                                    "type" : "integer"
+                                                  }
+                                                },
+                                                "required" : [ "firstNumber", "secondNumber" ]
+                                              }
+                                            }
+                                          } ],
+                                          "options" : {
+                                            "temperature" : 0.8,
+                                            "top_k" : 40,
+                                            "top_p" : 0.9
+                                          },
+                                          "stream" : true
+                                        }
+                                        """))
+
+                                .willReturn(aResponse()
+                                        .withHeader("Content-Type", "application/x-ndjson")
+                                        .withBody(
+                                                """
+                                                        {"model":"llama3.1","created_at":"2024-11-30T16:36:02.833930413Z","message":{"role":"assistant","content":"","tool_calls":[{"function":{"name":"sum","arguments":{"firstNumber":1,"secondNumber":1}}}]},"done":false}
+                                                        {"model":"llama3.1","created_at":"2024-11-30T16:36:04.368016152Z","message":{"role":"assistant","content":""},"done_reason":"stop","done":true,"total_duration":28825672145,"load_duration":29961281,"prompt_eval_count":169,"prompt_eval_duration":3906000000,"eval_count":22,"eval_duration":24887000000}""")));
+
+        wiremock()
+                .register(
+                        post(urlEqualTo("/api/chat"))
+                                .inScenario("")
+                                .whenScenarioStateIs("TOOL_CALL")
+                                .willSetStateTo("AI_RESPONSE")
+                                .withRequestBody(equalToJson("""
+                                        {
+                                          "model" : "llama3.2",
+                                          "messages" : [ {
+                                            "role" : "user",
+                                            "content" : "1 + 1"
+                                          }, {
+                                            "role" : "assistant",
+                                            "tool_calls" : [ {
+                                              "function" : {
+                                                "name" : "sum",
+                                                "arguments" : {
+                                                  "firstNumber" : 1,
+                                                  "secondNumber" : 1
+                                                }
+                                              }
+                                            } ]
+                                          }, {
+                                            "role" : "tool",
+                                            "content" : "2"
+                                          } ],
+                                          "tools" : [ {
+                                            "type" : "function",
+                                            "function" : {
+                                              "name" : "sum",
+                                              "description" : "Execute the sum of two numbers",
+                                              "parameters" : {
+                                                "type" : "object",
+                                                "properties" : {
+                                                  "firstNumber" : {
+                                                    "type" : "integer"
+                                                  },
+                                                  "secondNumber" : {
+                                                    "type" : "integer"
+                                                  }
+                                                },
+                                                "required" : [ "firstNumber", "secondNumber" ]
+                                              }
+                                            }
+                                          } ],
+                                          "options" : {
+                                            "temperature" : 0.8,
+                                            "top_k" : 40,
+                                            "top_p" : 0.9
+                                          },
+                                          "stream" : true
+                                        }
+                                        """))
+                                .willReturn(aResponse()
+                                        .withHeader("Content-Type", "application/x-ndjson")
+                                        .withBody(
+                                                """
+                                                        {"model":"llama3.1","created_at":"2024-11-30T16:36:04.368016152Z","message":{"role":"assistant","content":"The result is 2"},"done_reason":"stop","done":true,"total_duration":28825672145,"load_duration":29961281,"prompt_eval_count":169,"prompt_eval_duration":3906000000,"eval_count":22,"eval_duration":24887000000}""")));
+
+        var result = aiServiceWithTool.streaming("2", "1 + 1").collect().asList().await().indefinitely();
+        assertEquals(List.of("The result is 2"), result);
+
+        var messages = memory.getMessages("2");
+        assertEquals("1 + 1", ((dev.langchain4j.data.message.UserMessage) messages.get(0)).singleText());
+        assertEquals("The result is 2", ((dev.langchain4j.data.message.AiMessage) messages.get(3)).text());
+
+        if (messages.get(1) instanceof AiMessage aiMessage) {
+            assertTrue(aiMessage.hasToolExecutionRequests());
+            assertEquals("{\"firstNumber\":1,\"secondNumber\":1}", aiMessage.toolExecutionRequests().get(0).arguments());
+        } else {
+            fail("The second message is not of type AiMessage");
+        }
+
+        if (messages.get(2) instanceof ToolExecutionResultMessage toolResultMessage) {
+            assertEquals(2, Integer.parseInt(toolResultMessage.text()));
+        } else {
+            fail("The third message is not of type ToolExecutionResultMessage");
+        }
+    }
+}

--- a/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaChatLanguageModel.java
+++ b/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaChatLanguageModel.java
@@ -6,15 +6,12 @@ import static io.quarkiverse.langchain4j.ollama.MessageMapper.toOllamaMessages;
 import static io.quarkiverse.langchain4j.ollama.MessageMapper.toTools;
 
 import java.time.Duration;
-import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.jboss.logging.Logger;
-
-import com.fasterxml.jackson.core.JsonProcessingException;
 
 import dev.langchain4j.agent.tool.ToolExecutionRequest;
 import dev.langchain4j.agent.tool.ToolSpecification;
@@ -29,7 +26,6 @@ import dev.langchain4j.model.chat.listener.ChatModelResponse;
 import dev.langchain4j.model.chat.listener.ChatModelResponseContext;
 import dev.langchain4j.model.output.Response;
 import dev.langchain4j.model.output.TokenUsage;
-import io.quarkiverse.langchain4j.QuarkusJsonCodecFactory;
 
 public class OllamaChatLanguageModel implements ChatLanguageModel {
 
@@ -137,25 +133,10 @@ public class OllamaChatLanguageModel implements ChatLanguageModel {
                     AiMessage.from(response.message().content()),
                     new TokenUsage(response.promptEvalCount(), response.evalCount()));
         } else {
-            try {
-                List<ToolExecutionRequest> toolExecutionRequests = new ArrayList<>(toolCalls.size());
-                for (ToolCall toolCall : toolCalls) {
-                    ToolCall.FunctionCall functionCall = toolCall.function();
-
-                    // TODO: we need to update LangChain4j to make ToolExecutionRequest use a map instead of a String
-                    String argumentsStr = QuarkusJsonCodecFactory.ObjectMapperHolder.MAPPER
-                            .writeValueAsString(functionCall.arguments());
-                    toolExecutionRequests.add(ToolExecutionRequest.builder()
-                            .name(functionCall.name())
-                            .arguments(argumentsStr)
-                            .build());
-                }
-
-                result = Response.from(aiMessage(toolExecutionRequests),
-                        new TokenUsage(response.promptEvalCount(), response.evalCount()));
-            } catch (JsonProcessingException e) {
-                throw new RuntimeException("Unable to parse tool call response", e);
-            }
+            List<ToolExecutionRequest> toolExecutionRequests = toolCalls.stream().map(ToolCall::toToolExecutionRequest)
+                    .toList();
+            result = Response.from(aiMessage(toolExecutionRequests),
+                    new TokenUsage(response.promptEvalCount(), response.evalCount()));
         }
         return result;
     }

--- a/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaStreamingChatLanguageModel.java
+++ b/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaStreamingChatLanguageModel.java
@@ -2,23 +2,31 @@ package io.quarkiverse.langchain4j.ollama;
 
 import static dev.langchain4j.internal.ValidationUtils.ensureNotEmpty;
 import static io.quarkiverse.langchain4j.ollama.MessageMapper.toOllamaMessages;
+import static io.quarkiverse.langchain4j.ollama.MessageMapper.toTools;
 
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
 
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import dev.langchain4j.agent.tool.ToolSpecification;
 import dev.langchain4j.data.message.AiMessage;
 import dev.langchain4j.data.message.ChatMessage;
 import dev.langchain4j.model.StreamingResponseHandler;
 import dev.langchain4j.model.chat.StreamingChatLanguageModel;
 import dev.langchain4j.model.output.Response;
+import dev.langchain4j.model.output.TokenUsage;
 import io.smallrye.mutiny.Context;
 
 /**
  * Use to have streaming feature on models used trough Ollama.
  */
 public class OllamaStreamingChatLanguageModel implements StreamingChatLanguageModel {
+    private static final String TOOLS_CONTEXT = "TOOLS";
+    private static final String TOKEN_USAGE_CONTEXT = "TOKEN_USAGE";
+    private static final String RESPONSE_CONTEXT = "RESPONSE";
     private final OllamaClient client;
     private final String model;
     private final String format;
@@ -37,18 +45,23 @@ public class OllamaStreamingChatLanguageModel implements StreamingChatLanguageMo
     }
 
     @Override
-    public void generate(List<ChatMessage> messages, StreamingResponseHandler<AiMessage> handler) {
+    public void generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications,
+            StreamingResponseHandler<AiMessage> handler) {
         ensureNotEmpty(messages, "messages");
+        var tools = (toolSpecifications != null && toolSpecifications.size() > 0) ? toTools(toolSpecifications) : null;
 
         ChatRequest request = ChatRequest.builder()
                 .model(model)
                 .messages(toOllamaMessages(messages))
                 .options(options)
                 .format(format)
+                .tools(tools)
                 .stream(true)
                 .build();
 
-        Context context = Context.of("response", new ArrayList<ChatResponse>());
+        Context context = Context.empty();
+        context.put(RESPONSE_CONTEXT, new ArrayList<ChatResponse>());
+        context.put(TOOLS_CONTEXT, new ArrayList<ToolExecutionRequest>());
 
         client.streamingChat(request)
                 .subscribe()
@@ -58,13 +71,31 @@ public class OllamaStreamingChatLanguageModel implements StreamingChatLanguageMo
                             @SuppressWarnings("unchecked")
                             public void accept(ChatResponse response) {
                                 try {
-                                    if ((response == null) || (response.message() == null)
-                                            || (response.message().content() == null)
-                                            || response.message().content().isEmpty()) {
+                                    if ((response == null) || (response.message() == null)) {
                                         return;
                                     }
-                                    ((List<ChatResponse>) context.get("response")).add(response);
-                                    handler.onNext(response.message().content());
+
+                                    if (response.message().toolCalls() != null) {
+                                        List<ToolExecutionRequest> toolContext = context.get(TOOLS_CONTEXT);
+                                        List<ToolCall> toolCalls = response.message().toolCalls();
+                                        toolCalls.stream()
+                                                .map(ToolCall::toToolExecutionRequest)
+                                                .forEach(toolContext::add);
+                                    }
+
+                                    if (!response.message().content().isEmpty()) {
+                                        ((List<ChatResponse>) context.get(RESPONSE_CONTEXT)).add(response);
+                                        handler.onNext(response.message().content());
+                                    }
+
+                                    if (response.done()) {
+                                        TokenUsage tokenUsage = new TokenUsage(
+                                                response.evalCount(),
+                                                response.promptEvalCount(),
+                                                response.evalCount() + response.promptEvalCount());
+                                        context.put(TOKEN_USAGE_CONTEXT, tokenUsage);
+                                    }
+
                                 } catch (Exception e) {
                                     handler.onError(e);
                                 }
@@ -78,17 +109,37 @@ public class OllamaStreamingChatLanguageModel implements StreamingChatLanguageMo
                         },
                         new Runnable() {
                             @Override
-                            @SuppressWarnings("unchecked")
                             public void run() {
-                                var list = ((List<ChatResponse>) context.get("response"));
-                                StringBuilder builder = new StringBuilder();
-                                for (ChatResponse response : list) {
-                                    builder.append(response.message().content());
+
+                                TokenUsage tokenUsage = context.get(TOKEN_USAGE_CONTEXT);
+                                List<ChatResponse> chatResponses = context.get(RESPONSE_CONTEXT);
+                                List<ToolExecutionRequest> toolExecutionRequests = context.get(TOOLS_CONTEXT);
+
+                                if (toolExecutionRequests.size() > 0) {
+                                    handler.onComplete(Response.from(AiMessage.from(toolExecutionRequests), tokenUsage));
+                                    return;
                                 }
-                                AiMessage message = new AiMessage(builder.toString());
-                                handler.onComplete(Response.from(message));
+
+                                String response = chatResponses.stream()
+                                        .map(ChatResponse::message)
+                                        .map(Message::content)
+                                        .collect(Collectors.joining());
+
+                                AiMessage message = new AiMessage(response);
+                                handler.onComplete(Response.from(message, tokenUsage));
                             }
                         });
+    }
+
+    @Override
+    public void generate(List<ChatMessage> messages, ToolSpecification toolSpecification,
+            StreamingResponseHandler<AiMessage> handler) {
+        generate(messages, List.of(toolSpecification), handler);
+    }
+
+    @Override
+    public void generate(List<ChatMessage> messages, StreamingResponseHandler<AiMessage> handler) {
+        generate(messages, List.of(), handler);
     }
 
     /**

--- a/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/ToolCall.java
+++ b/model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/ToolCall.java
@@ -2,10 +2,27 @@ package io.quarkiverse.langchain4j.ollama;
 
 import java.util.Map;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import io.quarkiverse.langchain4j.QuarkusJsonCodecFactory;
+
 public record ToolCall(FunctionCall function) {
 
     public static ToolCall fromFunctionCall(String name, Map<String, Object> arguments) {
         return new ToolCall(new FunctionCall(name, arguments));
+    }
+
+    public ToolExecutionRequest toToolExecutionRequest() {
+        try {
+            return ToolExecutionRequest.builder()
+                    .name(function.name)
+                    .arguments(QuarkusJsonCodecFactory.ObjectMapperHolder.MAPPER
+                            .writeValueAsString(function.arguments()))
+                    .build();
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException("Unable to parse tool call response", e);
+        }
     }
 
     public record FunctionCall(String name, Map<String, Object> arguments) {

--- a/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/bean/TextChatMessage.java
+++ b/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/bean/TextChatMessage.java
@@ -1,6 +1,7 @@
 package io.quarkiverse.langchain4j.watsonx.bean;
 
 import static io.quarkiverse.langchain4j.watsonx.WatsonxUtils.base64Image;
+import static java.util.Objects.isNull;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -131,8 +132,9 @@ public sealed interface TextChatMessage
                     }
                     case IMAGE -> {
                         var imageContent = ImageContent.class.cast(content);
-                        var base64 = "data:image/%s;base64,%s".formatted(
-                                imageContent.image().mimeType(),
+                        var mimeType = imageContent.image().mimeType();
+                        var base64 = "data:%s;base64,%s".formatted(
+                                isNull(mimeType) ? "image" : mimeType,
                                 base64Image(imageContent.image()));
                         values.add(Map.of(
                                 "type", "image_url",
@@ -225,11 +227,8 @@ public sealed interface TextChatMessage
          */
         public static TextChatParameterTool of(ToolSpecification toolSpecification) {
             var toolParams = JsonSchemaElementHelper.toMap(toolSpecification.parameters());
-
-            var parameters = new TextChatParameterFunction(toolSpecification.name(), toolSpecification.description(), Map.of(
-                    "type", toolParams.get("type"),
-                    "properties", toolParams.get("properties"),
-                    "required", toolParams.get("required")));
+            var parameters = new TextChatParameterFunction(toolSpecification.name(), toolSpecification.description(),
+                    toolParams);
             return new TextChatParameterTool("function", parameters);
         }
     }

--- a/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/client/WatsonxRestApi.java
+++ b/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/client/WatsonxRestApi.java
@@ -142,6 +142,8 @@ public interface WatsonxRestApi {
 
         private static final Logger log = Logger.getLogger(WatsonClientLogger.class);
         private static final Pattern BEARER_PATTERN = Pattern.compile("(Bearer\\s*)(\\w{4})(\\w+)(\\w{4})");
+        private static final Pattern BASE64_IMAGE_PATTERN = Pattern.compile("(data:.+;base64,)(.{15})(.+)(.{15})([\\s\\S]*)");
+
         private final boolean logRequests;
         private final boolean logResponses;
 
@@ -205,7 +207,7 @@ public interface WatsonxRestApi {
             if (body == null) {
                 return "";
             }
-            return body.toString();
+            return formatBase64ImageForLogging(body.toString());
         }
 
         private String inOneLine(MultiMap headers) {
@@ -237,6 +239,26 @@ public interface WatsonxRestApi {
                 return sb.toString();
             } catch (Exception e) {
                 return "Failed to mask the API key.";
+            }
+        }
+
+        private static String formatBase64ImageForLogging(String body) {
+            try {
+
+                if (body == null || body.isBlank())
+                    return body;
+
+                Matcher matcher = BASE64_IMAGE_PATTERN.matcher(body);
+
+                StringBuilder sb = new StringBuilder();
+                while (matcher.find()) {
+                    matcher.appendReplacement(sb,
+                            matcher.group(1) + matcher.group(2) + "..." + matcher.group(4) + matcher.group(5));
+                }
+
+                return sb.isEmpty() ? body : sb.toString();
+            } catch (Exception e) {
+                return "Failed to format the base64 image value.";
             }
         }
 

--- a/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/client/filter/BearerTokenHeaderFactory.java
+++ b/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/client/filter/BearerTokenHeaderFactory.java
@@ -13,7 +13,7 @@ import io.smallrye.mutiny.Uni;
  */
 public class BearerTokenHeaderFactory extends ReactiveClientHeadersFactory {
 
-    private WatsonxTokenGenerator tokenGenerator;
+    private final WatsonxTokenGenerator tokenGenerator;
 
     public BearerTokenHeaderFactory(WatsonxTokenGenerator tokenGenerator) {
         this.tokenGenerator = tokenGenerator;

--- a/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/runtime/WatsonxRecorder.java
+++ b/model-providers/watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/runtime/WatsonxRecorder.java
@@ -156,10 +156,7 @@ public class WatsonxRecorder {
                 throw new ConfigValidationException(configProblems.toArray(EMPTY_PROBLEMS));
             }
 
-            String iamUrl = watsonConfig.iam().baseUrl().toExternalForm();
-            WatsonxTokenGenerator tokenGenerator = tokenGeneratorCache.computeIfAbsent(iamUrl,
-                    createTokenGenerator(watsonConfig.iam(),
-                            firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey())));
+            String apiKey = firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey());
 
             URL url;
             try {
@@ -170,7 +167,7 @@ public class WatsonxRecorder {
 
             EmbeddingModelConfig embeddingModelConfig = watsonConfig.embeddingModel();
             var builder = WatsonxEmbeddingModel.builder()
-                    .tokenGenerator(tokenGenerator)
+                    .tokenGenerator(createTokenGenerator(watsonConfig.iam(), apiKey))
                     .url(url)
                     .timeout(watsonConfig.timeout().orElse(Duration.ofSeconds(10)))
                     .logRequests(firstOrDefault(false, embeddingModelConfig.logRequests(), watsonConfig.logRequests()))
@@ -209,10 +206,7 @@ public class WatsonxRecorder {
             throw new ConfigValidationException(configProblems.toArray(EMPTY_PROBLEMS));
         }
 
-        String iamUrl = watsonConfig.iam().baseUrl().toExternalForm();
-        WatsonxTokenGenerator tokenGenerator = tokenGeneratorCache.computeIfAbsent(iamUrl,
-                createTokenGenerator(watsonConfig.iam(),
-                        firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey())));
+        String apiKey = firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey());
 
         URL url;
         try {
@@ -223,7 +217,7 @@ public class WatsonxRecorder {
 
         ScoringModelConfig rerankModelConfig = watsonConfig.scoringModel();
         var builder = WatsonxScoringModel.builder()
-                .tokenGenerator(tokenGenerator)
+                .tokenGenerator(createTokenGenerator(watsonConfig.iam(), apiKey))
                 .url(url)
                 .timeout(watsonConfig.timeout().orElse(Duration.ofSeconds(10)))
                 .logRequests(firstOrDefault(false, rerankModelConfig.logRequests(), watsonConfig.logRequests()))
@@ -242,17 +236,6 @@ public class WatsonxRecorder {
         };
     }
 
-    private Function<? super String, ? extends WatsonxTokenGenerator> createTokenGenerator(IAMConfig iamConfig, String apiKey) {
-        return new Function<String, WatsonxTokenGenerator>() {
-
-            @Override
-            public WatsonxTokenGenerator apply(String iamUrl) {
-                return new WatsonxTokenGenerator(iamConfig.baseUrl(), iamConfig.timeout().orElse(Duration.ofSeconds(10)),
-                        iamConfig.grantType(), apiKey);
-            }
-        };
-    }
-
     private WatsonxChatModel.Builder chatBuilder(LangChain4jWatsonxConfig runtimeConfig, String configName) {
         LangChain4jWatsonxConfig.WatsonConfig watsonConfig = correspondingWatsonRuntimeConfig(runtimeConfig, configName);
 
@@ -263,10 +246,7 @@ public class WatsonxRecorder {
         }
 
         ChatModelConfig chatModelConfig = watsonConfig.chatModel();
-        String iamUrl = watsonConfig.iam().baseUrl().toExternalForm();
-        WatsonxTokenGenerator tokenGenerator = tokenGeneratorCache.computeIfAbsent(iamUrl,
-                createTokenGenerator(watsonConfig.iam(),
-                        firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey())));
+        String apiKey = firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey());
 
         URL url;
         try {
@@ -276,7 +256,7 @@ public class WatsonxRecorder {
         }
 
         return WatsonxChatModel.builder()
-                .tokenGenerator(tokenGenerator)
+                .tokenGenerator(createTokenGenerator(watsonConfig.iam(), apiKey))
                 .url(url)
                 .timeout(watsonConfig.timeout().orElse(Duration.ofSeconds(10)))
                 .logRequests(firstOrDefault(false, chatModelConfig.logRequests(), watsonConfig.logRequests()))
@@ -306,10 +286,7 @@ public class WatsonxRecorder {
         }
 
         GenerationModelConfig generationModelConfig = watsonConfig.generationModel();
-        String iamUrl = watsonConfig.iam().baseUrl().toExternalForm();
-        WatsonxTokenGenerator tokenGenerator = tokenGeneratorCache.computeIfAbsent(iamUrl,
-                createTokenGenerator(watsonConfig.iam(),
-                        firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey())));
+        String apiKey = firstOrDefault(null, watsonConfig.apiKey(), runtimeConfig.defaultConfig().apiKey());
 
         URL url;
         try {
@@ -323,7 +300,7 @@ public class WatsonxRecorder {
         String promptJoiner = generationModelConfig.promptJoiner();
 
         return WatsonxGenerationModel.builder()
-                .tokenGenerator(tokenGenerator)
+                .tokenGenerator(createTokenGenerator(watsonConfig.iam(), apiKey))
                 .url(url)
                 .timeout(watsonConfig.timeout().orElse(Duration.ofSeconds(10)))
                 .logRequests(firstOrDefault(false, generationModelConfig.logRequests(), watsonConfig.logRequests()))
@@ -346,6 +323,18 @@ public class WatsonxRecorder {
                 .truncateInputTokens(generationModelConfig.truncateInputTokens().orElse(null))
                 .includeStopSequence(generationModelConfig.includeStopSequence().orElse(null))
                 .promptJoiner(promptJoiner);
+    }
+
+    private WatsonxTokenGenerator createTokenGenerator(IAMConfig iamConfig, String apiKey) {
+        return tokenGeneratorCache.computeIfAbsent(apiKey,
+                new Function<String, WatsonxTokenGenerator>() {
+                    @Override
+                    public WatsonxTokenGenerator apply(String iamUrl) {
+                        return new WatsonxTokenGenerator(iamConfig.baseUrl(),
+                                iamConfig.timeout().orElse(Duration.ofSeconds(10)),
+                                iamConfig.grantType(), apiKey);
+                    }
+                });
     }
 
     private LangChain4jWatsonxConfig.WatsonConfig correspondingWatsonRuntimeConfig(LangChain4jWatsonxConfig runtimeConfig,

